src/services/blocks/trace/types.ts

```diff
@@ -0,0 +1,203 @@
+import { AccountInfo, Address } from '@polkadot/types/interfaces';
+import BN from 'bn.js';
+
+export interface StringValues {
+	key?: string;
+	method?: 'Put' | 'Get';
+	result?: string;
+	[i: string]: unknown;
+}
+
+export interface Data {
+	stringValues: StringValues;
+}
+
+export interface TraceEvent {
+	name: string;
+	parentId: number;
+	target: string;
+	data: Data;
+}
+
+/**
+ * Parsed event.
+ */
+export interface PEvent extends TraceEvent {
+	storagePath: KeyInfo;
+	parentSpanId?: ParentSpanId;
+	eventIndex: number;
+}
+
+/**
+ * Parsed event with info about the action group it belongs to.
+ */
+export interface PActionEvent extends PEvent {
+	phase: PhaseId;
+	primarySpanId: ParentSpanId;
+}
+
+/**
+ * Parsed event of a `Get` from the `AccountInfo` storage item. Has info about
+ * the action group it belongs to and the decoded `AccountInfo` storage item.
+ */
+export interface PAccountEvent extends PActionEvent {
+	accountInfo: AccountInfo;
+	address: Address;
+}
+
+export interface TraceSpan {
+	id: number;
+	name: string;
+	parentId: number;
+	target: string;
+	data: Data;
+}
+
+/**
+ * Span with direct descendants
+ */
+export interface SpanWithChildren extends TraceSpan {
+	/**
+	 * Id of the child spans.
+	 */
+	children: number[];
+}
+
+export interface BlockTrace {
+	blockHash: string;
+	tracingTargets: string;
+	events: TraceEvent[];
+	spans: TraceSpan[];
+}
+
+export function isBlockTrace(
+	thing: unknown
+): thing is { blockTrace: BlockTrace } {
+	return (
+		typeof (thing as { blockTrace: BlockTrace })?.blockTrace.blockHash ===
+			'string' &&
+		typeof (thing as { blockTrace: BlockTrace })?.blockTrace.tracingTargets ===
+			'string' &&
+		Array.isArray((thing as { blockTrace: BlockTrace })?.blockTrace.events) &&
+		Array.isArray((thing as { blockTrace: BlockTrace })?.blockTrace.spans)
+	);
+}
+
+export interface TraceError {
+	error: string;
+}
+
+export function isTraceError(
+	thing: unknown
+): thing is { traceError: TraceError } {
+	return (
+		typeof (thing as { traceError: TraceError })?.traceError?.error === 'string'
```

Comment:
At first I was confused by this, and the return type of the traces call, but now it semi makes sense. The one thing that still confuses me a little is why would the `traceError` be undefined if it was returned in the object instead of `blockTrace`. Is this just a nuance of the type returned from polkadot-js?

Reply:
The return type is an enum - it can be one of two variants:

```rust
pub enum TraceBlockResponse {
	/// Error block tracing response
	TraceError(TraceError),
	/// Successful block tracing response
	BlockTrace(BlockTrace)
}
```
see [here](https://github.com/paritytech/substrate/pull/7780/files?file-filters%5B%5D=.lock&file-filters%5B%5D=.md&file-filters%5B%5D=.rs#diff-ab95a09c45453652418e5ecb30473d0fbbef66243b74bdd3486d56f37b164847R93-R98)
Because we are keying in deeply to an `unknown`, I figure we have better runtime safety if we check at every depth of the object for `undefined`